src/metldata/builtin_transformations/count_content_values/data_transform.py

```diff
@@ -0,0 +1,132 @@
+# Copyright 2021 - 2024 Universität Tübingen, DKFZ, EMBL, and Universität zu Köln
+# for the German Human Genome-Phenome Archive (GHGA)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Data transformation logic for count content values transformation."""
+
+from collections import Counter
+
+from schemapack.spec.datapack import DataPack
+
+from metldata.builtin_transformations.add_content_properties.path import (
+    resolve_data_object_path,
+)
+from metldata.builtin_transformations.common.path.path_utils import (
+    get_directly_referenced_class,
+)
+from metldata.builtin_transformations.count_content_values.instruction import (
+    CountContentValueInstruction,
+)
+from metldata.transform.exceptions import (
+    EvitableTransformationError,
+)
+
+
+class TransformationContext:
+    """Context manager for a data transformation process, wrapping the Datapack to be
+    transformed and transformation instructions per class.
+    """
+
+    def __init__(
+        self,
+        data: DataPack,
+        instructions_by_class: dict[str, list[CountContentValueInstruction]],
+    ):
+        self.data = data.model_copy(deep=True)
+        self.instructions_by_class = instructions_by_class
+
+    def get_class_resources(self, class_name: str):
+        """Extract the resources of a given class from a Datapack."""
+        resources = self.data.resources.get(class_name)
+        if not resources:
+            raise EvitableTransformationError()
+        return resources
+
+
+def count_content(
+    *,
+    data: DataPack,
+    instructions_by_class: dict[str, list[CountContentValueInstruction]],
+) -> DataPack:
+    """Transforms the data."""
+    context = TransformationContext(data, instructions_by_class)
+    for class_name, instructions in context.instructions_by_class.items():
+        resources = context.get_class_resources(class_name)
+
+        for instruction in instructions:
+            relation_path = instruction.source.relation_path
+            referenced_class = get_directly_referenced_class(relation_path)
+
+            # Only one element is expected in the path
+            relation_name = relation_path.elements[0].property
```

Comment:
This will raise in `get_directly_referenced_class` if the assumption is violated, right?